crates/compiler/mono/src/code_gen_help/refcount.rs

```diff
@@ -1501,7 +1511,7 @@ fn refcount_union_tailrec<'a>(
                     let mut tail_stmt = None;
                     for (i, field) in field_layouts.iter().enumerate() {
                         if i != tailrec_index {
-                            filtered.push(*field);
+                            filtered.push((i, *field));
```

Comment:
there was a bug that if the tailrec field was the 0th field, it would be decremented twice. We must keep track of the index of the fields that we still need to decrement.